mlir/include/mlir/Dialect/Linalg/Transforms/Transforms.h

```diff
@@ -1692,6 +1692,13 @@ void populateTransposeMatmulPatterns(RewritePatternSet &patterns,
 void populateBlockPackMatmulPatterns(RewritePatternSet &patterns,
                                      const ControlBlockPackMatmulFn &controlFn);
 
+/// Adds patterns that that reduce the rank of named contraction ops that have
+/// unit dimensions in the operand(s) by converting to a senquence of `collapse_shape`,
+/// `<corresponding linalg named op>`, `expand_shape` (if on tensors).  For example a
+/// `linalg.batch_matmul` with unit batch size will convert to `linalg.matmul`
+/// and a `linalg.matvec` with with unit spatial dim in lhs will convert to a `linalg.dot`.
+void populateContractionOpRankReducingPatterns(RewritePatternSet &patterns);
```

Comment:
I think it'd be useful to add a callback parameter to allow more fine-grained control.
For example, I might want to use this pattern to maximize number of `matmul` ops but would prefer to avoid going down to `matvec`.

Reply:
yah i was considering doing that. i was thinking of making it simple for now and just having a rank parameter that controls how low the rank is reduced to.  so if you want to only go down to matmul, rank would be 2. if you want to go down to matvec then rank 1 and dot rank 0.  would that work?

Reply:
I think sth simple like `std::function<bool(Operation *op)>` callback (just like `ControlPropagationFn` for example) should cover this and other use cases.

Reply:
A callback is more expensive both in runtime and in API complexity. If static rank is sufficient, let's stick with that until there is a real need in more fine-grained control.

Reply:
could this be a todo? i dont mind doing in this PR if it's insisted, but would be easy to extend in another PR

Reply:
Yeah, fair enough. No need to do scope creep.